R/remote.R

```diff
@@ -101,27 +101,24 @@ orderly_pull_archive <- function(name, id = "latest", root = NULL,
     orderly_log("pull", sprintf("%s already exists, skipping", label))
   } else {
     orderly_log("pull", label)
-    path <- file.path(config$root, "archive", name, id)
-    withCallingHandlers({
-      ## TODO (VIMC-2953): it would be might to have the remote pull
-      ## into a temporary directory and handle the copy ourselves for
-      ## easier rollback and less logic
-      remote$pull(name, id, config$root)
-
-      ## There's an assumption here that the depenency resolution here
-      ## will not be badly affected by migrations.  If a migration
-      ## changes how d$meta$depends is structured (if d$meta$depends
-      ## stops being a data.frame that includes name and id as
-      ## columns) then we'll need to deal with that when checking
-      ## dependencies too.
-      orderly_pull_resolve_dependencies(path, remote, config)
-
-      ## Only migrate after dependencies have been resolved because
-      ## some migrations do check dependencies.
-      migrate_single(path, config)
+    path <- remote$pull(name, id)
+
+    ## There's an assumption here that the depenency resolution here
+    ## will not be badly affected by migrations.  If a migration
+    ## changes how d$meta$depends is structured (if d$meta$depends
+    ## stops being a data.frame that includes name and id as
+    ## columns) then we'll need to deal with that when checking
+    ## dependencies too.
+    orderly_pull_resolve_dependencies(path, remote, config)
 
+    ## Only migrate after dependencies have been resolved because
+    ## some migrations do check dependencies.
+    migrate_single(path, config)
+
+    withCallingHandlers({
+      copy_directory(path, dest)
```

Comment:
How bad is the situation where the migrate succeeds but the copy_directory fails?
We end up with a report in the database but not in the archive.

I suppose if copy_directory fails there are bigger problems...

Reply:
I think the only way that this failure likely happens is that we're out of disk space (as we've already found that we can read from the source directory and that it contains a reasonable orderly report). Recovering would be a case of rebuilding the db.  This is hopefully uncommon enough that we can ignore it